### ISSUE TITLE
failover feature-gate Cannot be closed correctly

### ISSUE THREAD

**Issue by kubepopeye** [karmada-io/karmada#5375]
**Please provide an in-depth description of the question you have**:

I don't want karmada to trigger a failover when the cluster is unreachable, I tried to disable the feature-gate directly in karmada-controller and found that the failover still occurs!

**What do you think about this question?**:
I went to look at the karmada implementation, the cluster-controller does add a judgement for failover feature-gate in the monitor place, but in the `ttaintClusterByCondition` method, there is a lack of judgement, which leads to the taint being hit, and ultimately leads to feature-gate failover feature-gate.

**Environment**:
- Karmada version:
- v1.8.0
- Kubernetes version:
- 1.25
- Others:

**Comment by whitewindmills** [comment]
`taintClusterByCondition` only adds `NoSchedule` taints, which only affect scheduling.

**Comment by kubepopeye** [comment]
<img width="1802" alt="企业微信截图_ca74a2d1-ef4f-467e-8c83-00554c41388d" src="https://github.com/user-attachments/assets/c285253a-aa64-41e2-a216-9871fea0cd13">

**Comment by kubepopeye** [comment]
So what's causing this problem and can you help answer? It's true that it's noSchedule. but it's still triggering the clearing of the orphaned work.

@whitewindmills

**Comment by whitewindmills** [comment]
the orphan works may be caused by multiple reasons. I cannot find the root cause by those comments. can you paste scheduler logs here?

**Comment by kubepopeye** [comment]
scheduler logs？ I found that it seems to remove the resourcebind spec cluster only in the case of taint, which ends up causing findOrphan to be able to be found there.

**Comment by whitewindmills** [comment]
since you have disabled the failover feature, but karmada-scheduler might change their schduling result.

**Comment by kubepopeye** [comment]
> since you have disabled the failover feature, but karmada-scheduler might change their schduling result.

Is this the expected correct behaviour, I've found that in some cases it can lead to an empty list of Cluster's APIEnablements, which ends up with disastrous consequences!

**Comment by whitewindmills** [comment]
no, but we're fixing it.
ref:
https://github.com/karmada-io/karmada/pull/5325
https://github.com/karmada-io/karmada/pull/5216

**Comment by whitewindmills** [comment]
> I've found that in some cases it can lead to an empty list of Cluster's APIEnablements, which ends up with disastrous consequences!

have u ensured that's your root cause?

**Comment by kubepopeye** [comment]
> > I've found that in some cases it can lead to an empty list of Cluster's APIEnablements, which ends up with disastrous consequences!
> 
> have u ensured that's your root cause?

Yes, it seems to me that the shutdown of failover means that no migration of availability zones should take place, but it seems that there are some features here that cause failover-like behaviour to take place nonetheless.

**Comment by kubepopeye** [comment]
> > I've found that in some cases it can lead to an empty list of Cluster's APIEnablements, which ends up with disastrous consequences!
> 
> have u ensured that's your root cause?

In case of bad luck, the cluster-status-controller will empty the apiEnablements inside the cluster status when the cluster drops, and then the scheduler will intervene and find out that there is no API that matches, which will then lead to the cluster of the rb's spec to be emptied, and then finally the cluster departs. The binding controller's removeOrphan causes our downstream resources to be deleted. This is the complete chain, so we still think the failover implementation is incomplete.

Translated with www.DeepL.com/Translator (free version)

**Comment by whitewindmills** [comment]
first, there's nothing to do with FAILOVER.
did you see your cluster failing? it's important, wrong APIENABLEMENTS are usually caused by network errors or failing APIService.

todo:
grep such logs in your karmada-controller-manager.
* `Failed to get any APIs installed in Cluster`
* `Maybe get partial`
if you find them, that's the case.

**Comment by kubepopeye** [comment]
> * Failed to get any APIs installed in Cluster

yes,Failed to get any APIs installed in Cluster

<img width="1296" alt="企业微信截图_b69678f0-cbb8-4242-9a80-b414da0d8da9" src="https://github.com/user-attachments/assets/6fcb1462-fa4b-47cf-a500-b2fa7a02e58e">

**Comment by XiShanYongYe-Chang** [comment]
Hi @kubepopeye, thanks for your response, 

According to the log information, your analysis is correct, and we noticed this problem and fixed it in v1.12, just as @whitewindmills said, in terms of karmada-controller-manager, we [added CompleteAPIEnablements for cluster status](https://github.com/karmada-io/karmada/pull/5400). On the scheduler side, we [handled the cluster CompleteAPIEnablements Condition](https://github.com/karmada-io/karmada/pull/5216).

Now this problem should have been fixed, can you help make sure?

**Comment by XiShanYongYe-Chang** [comment]
Due to the long-term lack of activity, we plan to close this issue first. If you still encounter similar problems, please feel free to reopen this issue. Thanks to all of you!
/close
